### ISSUE TITLE
`torch.put_along_dim` and `torch.put_along_dim_` like `np.put_along_axis`

### ISSUE THREAD

**Issue by JasonGross** [pytorch/pytorch#120209]
### 🚀 The feature, motivation and pitch

Much like `torch.take_along_dim` was added as a wrapper around `torch.gather` with broadcasting logic in #52833 to match `np.take_along_axis`, it would be nice to have wrappers around `torch.scatter_` and `torch.scatter` that "are designed to work with [...] [f]unctions that return indices along a dimension, like [`torch.argmax()`](https://pytorch.org/docs/stable/generated/torch.argmax.html#torch.argmax) and [`torch.argsort()`](https://pytorch.org/docs/stable/generated/torch.argsort.html#torch.argsort)," as the [documentation for `take_along_dim`](https://pytorch.org/docs/stable/generated/torch.take_along_dim.html) indicates.

**Comment by albanD** [comment]
Is there a numpy equivalent function we want to mimick here?

**Comment by JasonGross** [comment]
Yes, I guess I mentioned it in the title but not in the body: [`np.put_along_axis`](https://numpy.org/doc/stable/reference/generated/numpy.put_along_axis.html)

**Comment by JasonGross** [comment]
I thin the wrapper is roughly just
```python
if dim is not None:
  return input.scatter(dim, index, src)
else:
  return input.flatten().scatter(0, index.flatten(), src.flatten()).reshape(input.shape)
```

**Comment by albanD** [comment]
Ho sorry missed that it was a put and not a take in the issue title haha

If there is a numpy equivalent we would definitely accept a PR that adds the same API yes!
That wrapper might even already exist in torch._numpy, cc @lezcano ?

**Comment by lezcano** [comment]
We didn't implement that one. That implementation sounds reasonable modulo missing checks on the inputs and explicit broadcasting (I believe `scatter` does not broadcast as much as `put_along_dim`).

A PR implementing this should make sure to get these two right.
